ballista/scheduler/src/scheduler_server/grpc.rs

```diff
@@ -124,14 +128,36 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> SchedulerGrpc
             };
 
             let mut tasks = vec![];
+            let mut prepare_failed_jobs = HashMap::<String, Vec<TaskDescription>>::new();
             for (_, task) in schedulable_tasks {
-                match self.state.task_manager.prepare_task_definition(task) {
+                let job_id = task.partition.job_id.clone();
+                if prepare_failed_jobs.contains_key(&job_id) {
+                    prepare_failed_jobs.entry(job_id).or_default().push(task);
+                    continue;
+                }
+                match self
+                    .state
+                    .task_manager
+                    .prepare_task_definition(task.clone())
+                {
                     Ok(task_definition) => tasks.push(task_definition),
                     Err(e) => {
                         error!("Error preparing task definition: {:?}", e);
+                        prepare_failed_jobs.entry(job_id).or_default().push(task);
                     }
                 }
             }
+
```

Comment:
I wonder whats the reason for this method? 

When we detect that preparation of task failed, we can not recover from it 
so job should be cancelled.

Would self.cancel_job(job_id) trigger cancelation of all running tasks for given job and do cleanup of execution graph?